### PR TITLE
[DO NOT MERGE] KAFKA-3705: non-key joins

### PR DESCRIPTION
**Pull request by guozhangwang**
This is just for reviewing the diff easily to see how it is done by @jfilipiak @Kaiserchen

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.java [hunk 1]
```diff
@@ -58,6 +58,14 @@
...
      */
     KeyValueIterator<K, V> all();
+    
+    /**
+     * Return an iterator over all keys that share the same prefix byte[] 
+     * as the provided key. 
+     * 
+     */
+    
+    KeyValueIterator<K, V> prefixScan(K prefix);
```
Comment (Kaiserchen) [inline streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.java:68]: @guozhangwang this is controversal and needs discussion. We use the same type for prefix and key. Currently I rely on the Serde to only render a prefix. that works fantastic with protobuf,avro and String might be the worst nightmare for JSON people

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 2]
```diff
@@ -1783,10 +1783,39 @@ void to(final Serde<K> keySerde,
...
+     * 							prefix of the output key wich is the same as serializing K 
+     * @param leftKeyExtractor a {@code ValueMapper} extracting the Key of this table from the resulting Key 
+     * @param <K0> the resultings tables Key
+     * @param <V0> the resultings tables Value
+     * @param joiner
+     * @return
+     */
+    <K0, V0, KO, VO> KTable<K0, V0> oneToManyJoin(KTable<KO, VO> other,
+			ValueMapper<VO, K> keyExtractor,
+			ValueMapper<K, K0> joinPrefixFaker,
```
Comment (Kaiserchen) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:1806]: Same controversy as with the Store prefix(K)

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 3]
```diff
@@ -1783,10 +1783,39 @@ void to(final Serde<K> keySerde,
...
+     * @param keyExtractor a {@code ValueMapper} returning the key of this table from the others value
+     * @param joinPrefixFaker a {@code ValueMapper} returning an outputkey that when serialized only produces the
+     * 							prefix of the output key wich is the same as serializing K 
+     * @param leftKeyExtractor a {@code ValueMapper} extracting the Key of this table from the resulting Key 
+     * @param <K0> the resultings tables Key
+     * @param <V0> the resultings tables Value
+     * @param joiner
+     * @return
+     */
+    <K0, V0, KO, VO> KTable<K0, V0> oneToManyJoin(KTable<KO, VO> other,
```
Comment (Kaiserchen) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:1804]: returns new K0 wich needs to contain K and K0. I love how this works so i expect controversy here aswell

**Comment by guozhangwang** [comment]
Closing this PR as it is already addressed by @bellemare

**Pull request closed without merging**